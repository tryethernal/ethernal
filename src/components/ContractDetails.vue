<template>
    <ImportArtifactModal ref="importModal" />
    <v-card>
        <v-card-text>
            <v-chip-group :selected-class="`text-${contrastingColor}`" v-model="activeTab" mandatory>
                <v-chip label size="small" value="code">Code</v-chip>
                <v-chip label size="small" value="read">Read Contract ({{ filteredCounts.read }})</v-chip>
                <v-chip label size="small" value="write">Write Contract ({{ filteredCounts.write }})</v-chip>
            </v-chip-group>

            <v-card v-if="userStore.isAdmin" class="mt-2">
                <v-card-text>
                    Manually override contract name & ABI <a href="#" class="text-primary text-decoration-none" @click.prevent="openImportModal">here</a>.
                </v-card-text>
            </v-card>

            <div v-show="activeTab === 'code'" class="mt-4">
                <ContractCode :contract="contract" />
            </div>

            <div v-show="activeTab === 'read' || activeTab === 'write'" class="mt-4">
                <ContractReadWrite 
                    :contract="contract" 
                    :force-tab="activeTab"
                    @update-filtered-counts="onUpdateFilteredCounts" />
            </div>
        </v-card-text>
    </v-card>
</template>

<script setup>
import { ref, computed, onMounted, watch } from 'vue';
import { useTheme } from 'vuetify';
import { useUserStore } from '../stores/user';
import { getBestContrastingColor } from '../lib/utils';
import ContractCode from './ContractCode.vue';
import ContractReadWrite from './ContractReadWrite.vue';
import ImportArtifactModal from './ImportArtifactModal.vue';

const props = defineProps(['contract']);

const userStore = useUserStore();
const activeTab = ref('code');
const filteredCounts = ref({ read: 0, write: 0 });
const importModal = ref(null);

// Add for chip group styling
const contrastingColor = computed(() => {
    const theme = useTheme();
    return getBestContrastingColor(theme.current.value.colors['surface-variant'], theme.current.value.colors);
});

const onUpdateFilteredCounts = (counts) => {
    filteredCounts.value = counts;
};

const updateTabFromHash = () => {
    const hash = window.location.hash.substring(1);
    if (hash === 'readContract') {
        activeTab.value = 'read';
    } else if (hash === 'writeContract') {
        activeTab.value = 'write';
    } else if (hash === 'code') {
        activeTab.value = 'code';
    }
};

const openImportModal = async () => {
    if (importModal.value) {
        const updated = await importModal.value.open({
            address: props.contract.address,
            name: props.contract.name,
            abi: props.contract.abi ? JSON.stringify(props.contract.abi, null, 2) : ''
        });

        if (updated)
            window.location.reload();
    }
};

// Watch for hash changes
onMounted(() => {
    updateTabFromHash();
    window.addEventListener('hashchange', updateTabFromHash);
});

// Update hash when tab changes
watch(() => activeTab.value, (newTab) => {
    const hashMap = {
        'code': 'code',
        'read': 'readContract',
        'write': 'writeContract'
    };
    if (window.location.hash.substring(1) !== hashMap[newTab]) {
        window.location.hash = hashMap[newTab];
    }
});
</script>

<style>
:deep(.v-chip-group) {
    overflow: visible !important;
}

:deep(.v-chip-group__content) {
    overflow: visible !important;
}
</style>
