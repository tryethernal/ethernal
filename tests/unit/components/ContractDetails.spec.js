import ContractDetails from '@/components/ContractDetails.vue';

const stubs = ['ContractCode', 'ContractReadWrite', 'ImportArtifactModal'];

describe('ContractDetails.vue', () => {
    const mockContract = {
        address: '0x123',
        name: 'TestContract',
        abi: JSON.stringify({ test: 'abi' })
    };

    beforeEach(() => {
        // Mock window.location.hash
        Object.defineProperty(window, 'location', {
            value: {
                hash: ''
            },
            writable: true
        });
    });

    it('Should show the component with code tab by default', async () => {
        const wrapper = mount(ContractDetails, {
            props: {
                contract: mockContract
            },
            global: {
                stubs,
                plugins: [
                    createTestingPinia({
                        initialState: {
                            user: { isAdmin: false }
                        }
                    })
                ]
            }
        });

        // Verify code tab is active
        const codeTab = wrapper.find('.v-chip-group .v-chip:first-child');
        expect(codeTab.classes()).toContain('text-primary');

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show read contract tab when hash is readContract', async () => {
        window.location.hash = '#readContract';
        
        const wrapper = mount(ContractDetails, {
            props: {
                contract: mockContract
            },
            global: {
                stubs,
                plugins: [
                    createTestingPinia({
                        initialState: {
                            user: { isAdmin: false }
                        }
                    })
                ]
            }
        });

        await wrapper.vm.$nextTick();

        // Verify read tab is active
        const readTab = wrapper.find('.v-chip-group .v-chip:nth-child(2)');
        expect(readTab.classes()).toContain('text-primary');

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show write contract tab when hash is writeContract', async () => {
        window.location.hash = '#writeContract';
        
        const wrapper = mount(ContractDetails, {
            props: {
                contract: mockContract
            },
            global: {
                stubs,
                plugins: [
                    createTestingPinia({
                        initialState: {
                            user: { isAdmin: false }
                        }
                    })
                ]
            }
        });

        await wrapper.vm.$nextTick();

        // Verify write tab is active
        const writeTab = wrapper.find('.v-chip-group .v-chip:nth-child(3)');
        expect(writeTab.classes()).toContain('text-primary');

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should update filtered counts when ContractReadWrite emits update', async () => {
        const wrapper = mount(ContractDetails, {
            props: {
                contract: mockContract
            },
            global: {
                stubs,
                plugins: [
                    createTestingPinia({
                        initialState: {
                            user: { isAdmin: false }
                        }
                    })
                ]
            }
        });

        await wrapper.findComponent({ name: 'ContractReadWrite' }).vm.$emit('update-filtered-counts', { read: 5, write: 3 });
        await wrapper.vm.$nextTick();

        // Verify counts are updated
        const readCount = wrapper.find('.v-chip-group .v-chip:nth-child(2)');
        const writeCount = wrapper.find('.v-chip-group .v-chip:nth-child(3)');
        expect(readCount.text()).toContain('Read Contract (5)');
        expect(writeCount.text()).toContain('Write Contract (3)');

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show import modal link for admin users', async () => {
        const wrapper = mount(ContractDetails, {
            props: {
                contract: mockContract
            },
            global: {
                stubs,
                plugins: [
                    createTestingPinia({
                        initialState: {
                            user: { isAdmin: true }
                        }
                    })
                ]
            }
        });

        // Verify import modal link is shown
        const importLink = wrapper.find('.v-card-text a');
        expect(importLink.exists()).toBe(true);
        expect(importLink.text()).toBe('here');

        expect(wrapper.html()).toMatchSnapshot();
    });
}); 