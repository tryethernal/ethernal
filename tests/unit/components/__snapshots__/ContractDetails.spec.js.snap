// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContractDetails.vue > Should show import modal link for admin users 1`] = `
"<import-artifact-modal-stub></import-artifact-modal-stub>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div class="v-slide-group v-chip-group v-theme--light" tabindex="0">
      <!---->
      <div class="v-slide-group__container">
        <div class="v-slide-group__content"><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal text-primary" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Code</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Read Contract (0)</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Write Contract (0)</div>
          <!---->
          <!----></span>
        </div>
      </div>
      <!---->
    </div>
    <div class="v-card v-theme--light v-card--density-default v-card--variant-elevated mt-2">
      <!---->
      <div class="v-card__loader">
        <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
          <!---->
          <div class="v-progress-linear__background"></div>
          <div class="v-progress-linear__buffer" style="width: 0%;"></div>
          <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
            <div class="v-progress-linear__indeterminate">
              <div class="v-progress-linear__indeterminate long"></div>
              <div class="v-progress-linear__indeterminate short"></div>
            </div>
          </transition-stub>
          <!---->
        </div>
      </div>
      <!---->
      <!---->
      <div class="v-card-text"> Manually override contract name &amp; ABI <a href="#" class="text-primary text-decoration-none">here</a>. </div>
      <!---->
      <!----><span class="v-card__underlay"></span>
    </div>
    <div class="mt-4">
      <contract-code-stub contract="[object Object]"></contract-code-stub>
    </div>
    <div class="mt-4" style="display: none;">
      <contract-read-write-stub contract="[object Object]" forcetab="code"></contract-read-write-stub>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;

exports[`ContractDetails.vue > Should show read contract tab when hash is readContract 1`] = `
"<import-artifact-modal-stub></import-artifact-modal-stub>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div class="v-slide-group v-chip-group v-theme--light" tabindex="0">
      <!---->
      <div class="v-slide-group__container">
        <div class="v-slide-group__content"><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Code</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal text-primary" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Read Contract (0)</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Write Contract (0)</div>
          <!---->
          <!----></span>
        </div>
      </div>
      <!---->
    </div>
    <!--v-if-->
    <div class="mt-4" style="display: none;">
      <contract-code-stub contract="[object Object]"></contract-code-stub>
    </div>
    <div class="mt-4" style="">
      <contract-read-write-stub contract="[object Object]" forcetab="read"></contract-read-write-stub>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;

exports[`ContractDetails.vue > Should show the component with code tab by default 1`] = `
"<import-artifact-modal-stub></import-artifact-modal-stub>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div class="v-slide-group v-chip-group v-theme--light" tabindex="0">
      <!---->
      <div class="v-slide-group__container">
        <div class="v-slide-group__content"><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal text-primary" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Code</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Read Contract (0)</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Write Contract (0)</div>
          <!---->
          <!----></span>
        </div>
      </div>
      <!---->
    </div>
    <!--v-if-->
    <div class="mt-4">
      <contract-code-stub contract="[object Object]"></contract-code-stub>
    </div>
    <div class="mt-4" style="display: none;">
      <contract-read-write-stub contract="[object Object]" forcetab="code"></contract-read-write-stub>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;

exports[`ContractDetails.vue > Should show write contract tab when hash is writeContract 1`] = `
"<import-artifact-modal-stub></import-artifact-modal-stub>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div class="v-slide-group v-chip-group v-theme--light" tabindex="0">
      <!---->
      <div class="v-slide-group__container">
        <div class="v-slide-group__content"><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Code</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Read Contract (0)</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal text-primary" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Write Contract (0)</div>
          <!---->
          <!----></span>
        </div>
      </div>
      <!---->
    </div>
    <!--v-if-->
    <div class="mt-4" style="display: none;">
      <contract-code-stub contract="[object Object]"></contract-code-stub>
    </div>
    <div class="mt-4" style="">
      <contract-read-write-stub contract="[object Object]" forcetab="write"></contract-read-write-stub>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;

exports[`ContractDetails.vue > Should update filtered counts when ContractReadWrite emits update 1`] = `
"<import-artifact-modal-stub></import-artifact-modal-stub>
<div class="v-card v-theme--light v-card--density-default v-card--variant-elevated">
  <!---->
  <div class="v-card__loader">
    <div class="v-progress-linear v-theme--light v-locale--is-ltr" style="top: 0px; height: 0px; --v-progress-linear-height: 2px;" role="progressbar" aria-hidden="true" aria-valuemin="0" aria-valuemax="100">
      <!---->
      <div class="v-progress-linear__background"></div>
      <div class="v-progress-linear__buffer" style="width: 0%;"></div>
      <transition-stub name="fade-transition" appear="false" persisted="false" css="true">
        <div class="v-progress-linear__indeterminate">
          <div class="v-progress-linear__indeterminate long"></div>
          <div class="v-progress-linear__indeterminate short"></div>
        </div>
      </transition-stub>
      <!---->
    </div>
  </div>
  <!---->
  <!---->
  <div class="v-card-text">
    <div class="v-slide-group v-chip-group v-theme--light" tabindex="0">
      <!---->
      <div class="v-slide-group__container">
        <div class="v-slide-group__content"><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal text-primary" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Code</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Read Contract (5)</div>
          <!---->
          <!----></span><span class="v-chip v-chip--label v-chip--link v-theme--light v-chip--density-default v-chip--size-small v-chip--variant-tonal" draggable="false" tabindex="0"><span class="v-chip__overlay"></span><span class="v-chip__underlay"></span>
          <!---->
          <!---->
          <div class="v-chip__content" data-no-activator="">Write Contract (3)</div>
          <!---->
          <!----></span>
        </div>
      </div>
      <!---->
    </div>
    <!--v-if-->
    <div class="mt-4">
      <contract-code-stub contract="[object Object]"></contract-code-stub>
    </div>
    <div class="mt-4" style="display: none;">
      <contract-read-write-stub contract="[object Object]" forcetab="code"></contract-read-write-stub>
    </div>
  </div>
  <!---->
  <!----><span class="v-card__underlay"></span>
</div>"
`;
